### PR TITLE
[build] add Android.bp support

### PR DESCRIPTION
**Pull request by zhanglongxia**
Since Android 7.0, the Soong build system was introduced in Android
to replace Make. This CL adds the Android.bp to support Soong build
system.

### CODE REVIEW

File: src/core/coap/coap.cpp [hunk 1]
```diff
@@ -654,6 +654,9 @@ void CoapBase::ProcessReceivedResponse(Message &aMessage, const Ip6::MessageInfo
     case kTypeConfirmable:
         // Send empty ACK if it is a CON message.
         IgnoreError(SendAck(aMessage, aMessageInfo));
+#ifdef __ANDROID__
+        [[clang::fallthrough]];
+#endif
```
Comment (wgtdkp) [inline src/core/coap/coap.cpp:659]: Is it possible to suppress the clang fall-through warning?

Comment (zhanglongxia) [reply src/core/coap/coap.cpp:659]: Yes, we can add `-Wno-implicit-fallthrough` to cppflags to suppress the clang fall-through warning. Should we suppress this warning? @jwhui @bukepo

Comment (wgtdkp) [reply src/core/coap/coap.cpp:659]: Not feeling good to have the platform-specific code everywhere :)

Another option is to have a macro `FALL_THROUGH` in `openthread/toolchain.h`:
```c++
#if __cplusplus >= 201703L
#define FALL_THROUGH    [[fallthrough]]
#if __CLANG__
#define FALL_THROUGH    [[clang::fallthrough]]
#elif __has_attribute(__fallthrough__)
# define FALL_THROUGH    __attribute__((__fallthrough__))
#else
# define FALL_THROUGH    do {} while (0)  /* fallthrough */
#endif
```

See https://elixir.bootlin.com/linux/v5.10/source/include/linux/compiler_attributes.h#L207

Comment (zhanglongxia) [reply src/core/coap/coap.cpp:659]: Will try it.

Comment (jwhui) [reply src/core/coap/coap.cpp:659]: Agree that we should avoid the `#if` statements.

Recognizing `// fall through` is common for existing toolchains, including gcc and coverity.

Comment (zhanglongxia) [reply src/core/coap/coap.cpp]: Updated

**Review by wgtdkp — APPROVED** [review]
LGTM :+1:

File: Android.mk [hunk 2]
```diff
@@ -145,6 +147,14 @@ LOCAL_CPPFLAGS                                                              := \
     -pedantic-errors                                                           \
     $(NULL)
 
+LOCAL_SHARED_LIBRARIES := libcutils
+
+ifeq ($(ANDROID_NDK),1)
+LOCAL_CFLAGS                                             += \
+    -DOPENTHREAD_ENABLE_ANDROID_NDK=1                       \
+    $(NULL)
+endif
```
Comment (bukepo) [inline Android.mk:156]:
```suggestion
ifeq ($(ANDROID_NDK),1)
LOCAL_SHARED_LIBRARIES += libcutils
LOCAL_CFLAGS                                             += \
    -DOPENTHREAD_ENABLE_ANDROID_NDK=1                       \
    $(NULL)
endif
```

**Review by bukepo — APPROVED** [review]
LGTM:+1:

File: script/check-android-build [hunk 3]
```diff
@@ -47,8 +47,8 @@ check_targets()
 
 main()
 {
-    USE_OTBR_DAEMON=1 check_targets ot-cli ot-ctl ot-ncp
-    check_targets ot-cli ot-ncp spi-hdlc-adapter
+    OPENTHREAD_ENABLE_ANDROID_MK=1 ANDROID_NDK=1 USE_OTBR_DAEMON=1 check_targets ot-cli ot-ctl ot-ncp
```
Comment (gjc13) [inline script/check-android-build:50]: Also add check without `ANDROID_NDK`?

**Review by gjc13 — APPROVED** [review]

File: Android.mk (hunk 2 above)
Comment (zhanglongxia) [reply Android.mk]: Updated

File: include/openthread-config-android-version.h.in [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
```
Comment (jwhui) [inline include/openthread-config-android-version.h.in:2]: nit: can we move this file to something like `etc/android`? Similar to what we did with `etc/cmake/openthread-config-generic.h.in`?

Comment (zhanglongxia) [reply include/openthread-config-android-version.h.in]: Updated

**Review by jwhui — APPROVED** [review]

**Pull request merged**